### ISSUE TITLE
Allow Duktape/C functions to initiate a tailcall

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#1026]
This is needed for ES6 compliance so it will be useful internally in any case.

**Comment by svaarala** [comment]
How would that work in practice? Would the Duktape/C function call an API call to indicate "setup for a tailcall" and then return so that the native callstack for the current call is unwound before executing the next?

**Comment by svaarala** [comment]
As an idiom maybe something like:

```c
return duk_tailcall(ctx, 3 /*nargs*/);
```

In practice `duk_tailcall()` wouldn't make the actual call but would just set up the tailcall. The return code would then be something to indicate to the *previous* call handling to complete the tailcall.

Unfortunately one cannot rely on the compiler doing an actual native tailcall there, so it would need to be something like a setup call followed by a return to trigger the actual call handling.

**Comment by fatcerberus** [comment]
FWIW, ES6 specifies it in terms of a setup operation too.  The second to last step is to PrepareForTailCall().

**Comment by svaarala** [comment]
That would mean that if you did a `duk_tailcall()` and continued working with the value stack all bets would be off (the call stack and value stack might already be partially unwound). That needs clear documentation and the idiom `return duk_tailcall()` would make it pretty clear what the intent was.

As far as implementation options, there are several alternatives for splitting the necessary before exiting the current C call vs. after exiting the current C call in the call handling of the previous call.

For example, `duk_tailcall()` could do all the necessary value stack setup but avoid unwinding the call stack just yet. A special return code (`DUK_RET_TAILCALLED`) would then be picked up by the call handling which would proceed to setup the call stack too. This would keep the call stack intact while we're still technically in the tailcalling C function.

Not that it really matters: doing anything with the Duktape context after the `duk_tailcall()` should be undefined behavior IMO, so it would be also fine to handle the call stack setup within `duk_tailcall()` too.

**Comment by fatcerberus** [comment]
ES6 Section 14.6.3 - PrepareForTailCall:
http://www.ecma-international.org/ecma-262/6.0/#sec-preparefortailcall

**Comment by svaarala** [comment]
I read that - it doesn't really provide much guidance for the practical implementation.

**Comment by svaarala** [comment]
To be more precise it specifies how things need to look like at a very high level. The specification needs to talk about the stack not growing etc, without assuming too much about the implementation because that varies a great deal.

But looking at a concrete implementation, there are many different implementation options to achieve the required effect, and also many C API designs (which is outside the scope of any ES6 requirements naturally).

For example, using a `duk_tailcall()` to do some setup (or just stash values) and using a magic return value to allow the call handling (of the current call) to deal with the tailcall is one option.

Another option is for `duk_tailcall()` to do some setup and `longjmp()` to implement the tailcall. This would mean a `setjmp()` for every call so I wouldn't want to go with this option.

For the `duk_tailcall()` + magic return option, it could also be a macro like:

```
#define duk_tailcall(ctx,nargs) do { \
        return duk_tailcall_setup((ctx), (nargs)); \
    } while (0)
```

In other words, the necessary return could be forced by the macro so that it wouldn't be possible to continue using any API calls accidentally.

**Comment by fatcerberus** [comment]
I like that macro a lot.  It doesn't need a lot of explaining in documentation either, other than to say "this function never returns".

Of course the downside to that is that it can't easily be converted into a real function later while still retaining the noreturn property unless you refactor it to use `longjmp()` (which is undesirable as you say).

**Comment by fatcerberus** [comment]
One workable option might be to allow the caller to push the function and arguments to be tailcalled onto the value stack and then return a value, perhaps computed by a macro, which encodes the argument count and related flags, which can be picked up during unwind.  Something like:

```c
duk_push_c_function(ctx, ...);
duk_push_string(ctx, "arg 1");
duk_push_int(ctx, 2);
return DUK_TAILCALL(2 /*nargs*/, 0 /*flags*/);
```

This seems cleaner and less error prone than having an "undefined behavior" clause.

**Comment by svaarala** [comment]
That's one alternative for the mechanics but I don't see how that significantly improves the API contract. It would still allow this:

```
duk_push_c_function(ctx, ...);
duk_push_string(ctx, "arg 1");
duk_push_int(ctx, 2);
DUK_TAILCALL(2 /*nargs*/, 0 /*flags*/);
duk_push_null();
return 1;
```

Which would set up for a tail call, then use the computed macro (ignoring the result), and then do something unrelated. This would be pretty confusing because the `DUK_TAILCALL()` macro would be a no-op but appears to be a function call.

**Comment by fatcerberus** [comment]
Hm, I guess that's true, but that might just be an issue of naming more than anything.  It could also be called `DUK_RET_TAILCALL()`, which then resembles the magic error return constants (DUK_RET_ERROR` et al.) enough that it might not be as easily confused.

I don't know, just throwing an idea out there that came to me. :)

In any case I don't really expect this issue to be solved in the short term, which is why I posted an issue to track it.  I personally thought the tailcall clauses in ES6 were kind of odd, although I guess it does let you do things like this without exhausting the stack:

```js
function factorial(n) {
    /* for the sake of argument pretend `this` is needed for recursive calls */
    return n * factorial.call(this, n - 1);  /* guaranteed tailcall */
}
```

**Comment by fatcerberus** [comment]
Here's another idea which might work better:
`duk_tailcall()` sets a flag on the activation record that says "I want to do a tailcall".  Once that flag is set the semantics of the return value change, so that instead of being the number of values returned it instead specifies e.g. the number of arguments to be passed to the tailcalled function.  From there you'd just push the function to call and its arguments and then return as usual.

In this way the tailcall flag can be set at any time and the function can safely work with the value stack until it actually returns.

Maybe in this case the setup function could be named e.g. `duk_prep_tailcall()` to underline the fact that it's only setting the stage instead of directly doing anything.  And that naming also matches the way tailcalls are described in ES6, which is nice.

**Comment by svaarala** [comment]
> I personally thought the tailcall clauses in ES6 were kind of odd, although I guess it does let you do things like this without exhausting the stack [...]

The specific requirements for ES6 tailcalls don't match the current implementation and it may be difficult to get them all to work with the current compiler - tailcall support has been implemented in a very, very simple way at present :-)

**Comment by fatcerberus** [comment]
Thinking about this now, I think this is another case where it'd probably be best to implement the Duktape/C tailcall support internally first, without exposing it to the public API.  This way we could experiment with different approaches to see what works best, as it was with value summarization.
